src/main/java/com/scalar/db/storage/jdbc/JDBC.java

```diff
@@ -0,0 +1,187 @@
+import com.scalar.db.api.Delete;
+import com.scalar.db.api.DistributedStorage;
+import com.scalar.db.api.Get;
+import com.scalar.db.api.Mutation;
+import com.scalar.db.api.Put;
+import com.scalar.db.api.Result;
+import com.scalar.db.api.Scan;
+import com.scalar.db.api.Scanner;
+import com.scalar.db.config.DatabaseConfig;
+import com.scalar.db.exception.storage.ExecutionException;
+import com.scalar.db.exception.storage.NoMutationException;
+import com.scalar.db.storage.jdbc.metadata.TableMetadataManager;
+import com.scalar.db.storage.jdbc.query.QueryBuilder;
+import org.apache.commons.dbcp2.BasicDataSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * A storage implementation with JDBC for {@link DistributedStorage}.
+ *
+ * <p>Note that the consistency in an operation is always LINEARIZABLE in this implementation. Even
+ * if consistency is specified in an operation, it will be ignored.
+ *
+ * @author Toshihiro Suzuki
+ */
+@ThreadSafe
+public class JDBC implements DistributedStorage {
+  private static final Logger LOGGER = LoggerFactory.getLogger(JDBC.class);
+
+  private final BasicDataSource dataSource;
+  private final JDBCService jdbcService;
+  @Nullable private String defaultSchema;
+  @Nullable private String defaultTable;
+
+  @Inject
+  public JDBC(DatabaseConfig config) {
+    dataSource = JDBCUtils.initDataSource(config);
+
+    String schemaPrefix = config.getNamespacePrefix().orElse("");
+
+    TableMetadataManager tableMetadataManager = new TableMetadataManager(dataSource, schemaPrefix);
+    OperationChecker operationChecker = new OperationChecker(tableMetadataManager);
+    QueryBuilder queryBuilder =
+        new QueryBuilder(
+            tableMetadataManager, JDBCUtils.getRDBType(config.getContactPoints().get(0)));
+    jdbcService = new JDBCService(operationChecker, queryBuilder, schemaPrefix);
+  }
+
+  @VisibleForTesting
+  JDBC(BasicDataSource dataSource, JDBCService jdbcService) {
+    this.dataSource = dataSource;
+    this.jdbcService = jdbcService;
+  }
+
+  @Override
+  public void with(String namespace, String tableName) {
+    defaultSchema = namespace;
+    defaultTable = tableName;
+  }
+
+  @Override
+  public void withNamespace(String namespace) {
+    defaultSchema = namespace;
+  }
+
+  @Override
+  public Optional<String> getNamespace() {
+    return Optional.ofNullable(defaultSchema);
+  }
+
+  @Override
+  public void withTable(String tableName) {
+    defaultTable = tableName;
+  }
+
+  @Override
+  public Optional<String> getTable() {
+    return Optional.ofNullable(defaultTable);
+  }
+
+  @Override
+  public Optional<Result> get(Get get) throws ExecutionException {
+    try (Connection connection = dataSource.getConnection()) {
```

Comment:
I think the following part is a bit redundant and is shown several times, so it would be great if it can be unified.

```
    try (Connection connection = dataSource.getConnection()) {
       // ...
    } catch (SQLException e) {
      throw new ExecutionException("An error occurred", e);
    }
```

I have't fully verified it, but how about having a method like this?

```
  @Override
  public Optional<Result> get(Get get) throws ExecutionException {
    ThrowableFunction<Get, Optional<Result>> f =
        o -> jdbcService.get(o, connection, defaultSchema, defaultTable);
    processOperation(f);
  }

 private void processOperation(ThrowableFunction f) {
    try (Connection connection = dataSource.getConnection()) {
      return f.apply();
    } catch (SQLException e) {
      throw new ExecutionException("An error occurred", e);
    }
  }

// it can be separately defined.
@FunctionalInterface
public interface ThrowableFunction<T, R> {
  R apply(T t) throws Exception;
}
```